lib/Secrets/secrets_c.c

```diff
@@ -0,0 +1,1065 @@
+#include "secrets_c.h"
+
+#include <dbus/dbus.h>
+#include <sys/types.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+int Sailfish_Secrets__getServerPeerToPeerAddress(
+        char ** p2pAddr,
+        DBusConnection **sessionBus)
+{
+    DBusMessage* msg;
+    DBusMessageIter args;
+    DBusError err;
+    DBusPendingCall* pending;
+    char *param = NULL;
+
+    /* initialise the error */
+    dbus_error_init(&err);
+
+    /* connect to the session bus */
+    *sessionBus = dbus_bus_get(DBUS_BUS_SESSION, &err);
+    if (dbus_error_is_set(&err)) {
+        fprintf(stderr, "Connection Error (%s)\n", err.message);
+        dbus_error_free(&err);
+    }
+    if (*sessionBus == NULL) {
+        return 0;
+    }
+
+    /* create a new method call and check for errors */
+    msg = dbus_message_new_method_call(
+                "org.sailfishos.secrets.daemon.discovery",
+                "/Sailfish/Secrets/Discovery",
+                "org.sailfishos.secrets.daemon.discovery",
+                "peerToPeerAddress");
+    if (msg == NULL) {
+        fprintf(stderr, "Message Null\n");
+        dbus_error_free(&err);
+        return 0;
+    }
+
+    /* send message and get a handle for a reply */
+    dbus_connection_send_with_reply(
+                *sessionBus,
+                msg,
+                &pending,
+                -1);
```

Comment:
Blocking call detected! And there are more all over the place. That's no good.

Reply:
Yes, the C API is currently purely synchronous.  Yes, we should add asynchronous API.  I just created this quick API as a proof of concept and something we can iterate on.